`@retroactive` conformance syntax and checking

harlanhaskins opened the pull request:
Updated with the Core Team's new recommendation.

Warns whenever a conformance declared in an extension is "retroactive", i.e. it declares a conformance of a type outside this module to a protocol also outside this module.

This warning can be silenced via `@retroactive`, like so:

```swift
extension Date: @retroactive Identifiable { ... }
```

This also disables the warning for the Foundation module, which has a bunch of retroactive conformances of stdlib types to `_ObjectiveCBridgeable`.

Fixes rdar://74543558

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
After feedback from @airspeedswift I'm moving from `@_retroactiveConformance` to requiring you to module-qualify everything from outside your module explicitly. This isn't quite done yet

CodaFi commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1401,6 +1402,188 @@ static void maybeDiagnoseClassWithoutInitializers(ClassDecl *classDecl) {
+  // If we're compiling an overlay and the type comes from the clang module,
+  // conformances aren't retroactive.
+  if (auto clangModule = extTypeModule->findUnderlyingClangModule()) {
+    if (clangModule->getTopLevelModuleName() == module->getName().str()) {
```
I wonder if we even need to go this far. I think all you need to do is something like

```
static bool isClangOverlayOf(ModuleDecl *potentialOverlay,
                             ModuleDecl *potentialUnderlying) {
  return !potentialOverlay->isNonSwiftModule() &&
         potentialUnderlying->isNonSwiftModule() &&
         potentialOverlay->getName() == potentialUnderlying->getName();
}
```

If you promoted this to API on `ModuleDecl` that'd be swell.

CodaFi commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1401,6 +1402,188 @@ static void maybeDiagnoseClassWithoutInitializers(ClassDecl *classDecl) {
+    return false;
+  }
+  
+  return components.front()->getNameRef().isSimpleName(module->getName());
```
@beccadax We'll have to remember to update this when module selectors go through.

beccadax replied:
Harlan, could you add a comment starting with `// FIXME(ModQual):` somewhere in this function so I don't miss it?

harlanhaskins replied:
Done, thanks! I'm excited to see that eventually get traction!

CodaFi commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1401,6 +1402,188 @@ static void maybeDiagnoseClassWithoutInitializers(ClassDecl *classDecl) {
+    // First, push the inherited protocols onto the worklist.
+    for (ProtocolDecl *inherited : protoDecl->getInheritedProtocols()) {
+      if (seen.insert(inherited).second) {
+        workList.push_back(inherited);
```
A slight optimization of this could be to break the loop after you've encountered a protocol you've already seen, since you know that its ancestors have all been pushed onto the worklist by this very loop.

CodaFi commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1401,6 +1402,188 @@ static void maybeDiagnoseClassWithoutInitializers(ClassDecl *classDecl) {
+    // If the protocol comes from the underlying clang module of the extension's
+    // module, skip it.
+    if (auto clangModule = protoModule->findUnderlyingClangModule()) {
+      if (clangModule->getTopLevelModuleName() == module->getName().str()) {
```
Same quip as above about adding something like `ModuleDecl::isOverlayOf`

CodaFi commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1401,6 +1402,188 @@ static void maybeDiagnoseClassWithoutInitializers(ClassDecl *classDecl) {
+  // extension declares a conformance to.
+
+  llvm::SmallPtrSet<ProtocolDecl *, 8> seen;
+  llvm::SmallSetVector<ProtocolDecl *, 8> externalProtocols;
```
Seems like the `seen` list is already giving you uniqueness here.

harlanhaskins replied:
I need to be able to quickly check containment of the external protocols, not every protocol that's been seen by the work list

CodaFi commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1401,6 +1402,188 @@ static void maybeDiagnoseClassWithoutInitializers(ClassDecl *classDecl) {
+    }
+  }
+
+  if (externalProtocols.empty()) {
```
This also occurs on lines [1529-1532](https://github.com/apple/swift/pull/36068/files#diff-fa1198ad7fbbed0b2ad70245eaa658121ebd138b3d92dcb8013d14ce906bfe58R1529-R1532)

harlanhaskins replied:
I will have removed an entry from this list on line 1562, but I'll add a comment mentioning that

harlanhaskins replied:
This code as-written is more clever than I'd like it to be

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test Windows

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test macOS

CodaFi commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1401,6 +1402,203 @@ static void maybeDiagnoseClassWithoutInitializers(ClassDecl *classDecl) {
+
+  // If the extended type name isn't qualified with the module name, add it.
+  if (!isModuleQualified(extTypeRepr, extTypeModule)) {
+    llvm::SmallString<12> qualifiedName;
```
`12` is kind of an odd choice here. Perhaps a power-of-two: 32 would be nice.

CodaFi commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1401,6 +1402,203 @@ static void maybeDiagnoseClassWithoutInitializers(ClassDecl *classDecl) {
+        dyn_cast_or_null<ProtocolDecl>(tyLoc.getType()->getAnyNominal());
+    if (protoDecl && externalProtocols.remove(protoDecl)) {
+      llvm::SmallString<12> qualifiedName;
+      llvm::raw_svector_ostream os(qualifiedName);
```
Same here

CodaFi commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1401,6 +1402,203 @@ static void maybeDiagnoseClassWithoutInitializers(ClassDecl *classDecl) {
+/// it's not supported to declare these conformances because any other framework
+/// (including the source framework) can declare this conformance, which would
+/// result in duplicate symbols at runtime.
+static void diagnoseRetroactiveConformances(ExtensionDecl *ext) {
```
Can we assert that the extension
1) Isn't NULL
2) Isn't deserialized.

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test Linux

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test Windows

harlanhaskins commented:
@swift-ci please smoke test Windows

varungandhi-apple commented:
Apart from the merge conflict, is this mostly in a good state/should we land this soon? (And maybe cherry-pick to 5.5?)

dabrahams commented:
Why is retroactive conformance a problem when the protocol is declared in the same module?

harlanhaskins commented:
It's not; this diagnostic explicitly only warns for conformances of types declared outside the current module to protocols declared outside the current module.

Though I realize looking at the test that it's missing a check that it doesn't diagnose.

I should bring this up on evolution now, I don't think there's any work left on it besides adding that test case. The core team felt this warning warranted an evolution proposal.

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
Oop, accidentally committed python -> python3 changes, will remove

harlanhaskins committed:
Warn when a resilient extension declares a retroactive conformance

It's not particularly _great_ when executable targets declare these, but
it's certainly not okay when a library that's meant to be distributed
declares these. When library evolution is enabled, warn about
retroactive conformances.

To silence the warning, add the @retroactive attribute to the extended type,
similarly to the @unchecked attribute for Sendable. It is an error to use
@retroactive for anything but a retroactive conformance declaration in
an extension.

Fixes rdar://74543558

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@beccadax @CodaFi @slavapestov Hi folks! I think this is ready for review post-evolution-acceptance, thanks!

beccadax commented on lib/AST/Module.cpp:
```diff
@@ -2685,6 +2685,38 @@ ModuleDecl::getDeclaringModuleAndBystander() {
+    return true;
+  }
+
+  if (getUnderlyingModuleIfOverlay() == other) {
+    return true;
+  }
+
+  if (this == other->getUnderlyingModuleIfOverlay()) {
+    return true;
+  }
```
Nit: Why not use your cute little helper method?

```suggestion
  if (this->isClangOverlayOf(other) || other->isClangOverlayOf(this)) {
    return true;
  }
```

harlanhaskins replied:
🤦 thanks for recognizing that!

beccadax commented on lib/AST/Module.cpp:
```diff
@@ -2685,6 +2685,38 @@ ModuleDecl::getDeclaringModuleAndBystander() {
+  // If the type has the @_originallyDefinedIn attribute, check if this is the
+  // "originally defined in" module.
+
+  if (getName().is(other->getAlternateModuleName())) {
+    return true;
+  }
+
+  if (other->getName().is(getAlternateModuleName())) {
+    return true;
+  }
```
Wait—`this` and `other` are both modules, right? Is `getAlternateModuleName()` ever a thing for `ModuleDecl`s? If not, do you need to do something else to handle `@_originallyDefinedIn`?

harlanhaskins replied:
Yeah, so I just realized this check wasn't doing anything, but the existing check is sufficient to cover `@_originallyDefinedIn` (and I added a test to verify that)

beccadax commented on test/Concurrency/sendable_preconcurrency.swift:
```diff
@@ -39,7 +39,7 @@ func testA(ns: NS, mt: MyType, mt2: MyType2, mt3: MyType3, sc: StrictClass, nsc:
 }
 
-extension NonStrictStruct: @unchecked Sendable { }
+extension NonStrictStruct: @retroactive @unchecked Swift.Sendable { }
```
Is the added `Swift.` here necessary?

xwu replied:
We are still permitting the previous silencing mechanism given that there won't be a way to use `@retroactive` in a way that preserves source compatibility with existing language versions, yes?

harlanhaskins replied:
Indeed; there's a test that covers silencing the warning via module qualification

beccadax commented on test/Sema/extension_retroactive_conformances.swift:
```diff
@@ -0,0 +1,73 @@
+extension Sample3: NestedInheritsSampleProtocol {} // expected-warning {{extension declares a conformance of imported type 'Sample3' to imported protocol 'SampleProtocol1'}}
+// expected-note @-1 {{add '@retroactive' to silence this warning}} {{1-1=extension Sample3: @retroactive SampleProtocol1 {\}\n}}
+
+extension Sample4: NestedInheritsMultipleSampleProtocols {} // expected-warning {{extension declares a conformance of imported type 'Sample4' to imported protocols 'SampleProtocol2', 'SampleProtocol1'}}
```
Wonderfully thorough; great work.

beccadax commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -315,6 +315,7 @@ PrintOptions PrintOptions::printSwiftInterfaceFile(ModuleDecl *ModuleToPrint,
     DAK_StaticInitializeObjCMetadata,
     DAK_RestatedObjCConformance,
     DAK_NonSendable,
+    TAK_retroactive
```
Wow, the capitalization difference between `DAK_`s and `TAK_`s is kinda disgusting, but that’s not your fault.😅

Should we really leave `@retroactive` out of module interfaces? The Steering Group discussed the possibility of eventually having privileged-but-separate modules, like cross-import overlays or modules in the same package group, where `@retroactive` might be optional and its presence would affect the ABI; if that happened, we would certainly want all retroactive conformances marked in module interfaces.

(If you’re merely worried about backwards compatibility with older compilers, we now have an infrastructure for detecting when a declaration uses a new feature and printing an alternate version of it into the module interface using `#if`.)

tshortli reviewed (commented):
Nice, really looking forward to having this implemented.

I gather that the intent is to not print `@retroactive` in `.swiftinterfaces` (makes sense). Could you add a `ModuleInterface` tests that asserts this explicitly?

harlanhaskins commented:
I'm actually working on an update here to print the attribute in module interfaces, guarded with a `#if hasAttribute` check, but it'll have to be something like

```
#if hasAttribute(retroactive)
extension Optional: @retroactive CustomDebugStringConvertible where Wrapped: CustomDebugStringConvertible {}
#else
extension Optional: CustomDebugStringConvertible where Wrapped: CustomDebugStringConvertible {}
#endif

extension Optional: where Wrapped: CustomDebugStringConvertible {
    public var description: String {
        // ...
    }
}
```

harlanhaskins commented:
@tshortli @beccadax Sorry for the extreme delay on this! I think I have a workable implementation now that prints `@retroactive` in module interfaces guarded with a `$RetroactiveAttribute` feature flag check using the standard feature compatibility printing path, and added a test to make sure of that.

Thank you! I think this PR is _good to go_.

harlanhaskins commented:
@swift-ci please smoke test

tshortli commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -342,7 +342,7 @@ PrintOptions PrintOptions::printSwiftInterfaceFile(ModuleDecl *ModuleToPrint,
     DAK_StaticInitializeObjCMetadata,
     DAK_RestatedObjCConformance,
-    DAK_NonSendable,
+    DAK_NonSendable
```
👀

tshortli commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1569,6 +1570,207 @@ static void maybeDiagnoseClassWithoutInitializers(ClassDecl *classDecl) {
+  }
+
+  // Don't warn for this if we see it in module interfaces. We shouldn't
+  // write out the `@retroactive` attribute in module interfaces.
```
This comment is out-of-date now right?

tshortli reviewed (approved):
LGTM! Is there a https://github.com/apple/swift-syntax PR as well?

harlanhaskins committed:
Address feedback, remove redundant test, and add test for @_originallyDefinedIn

harlanhaskins committed:
Remove unnecessary changes in Module

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
Is this good to merge? 🙇

tshortli commented:
🚢
